README.md
# lvslb-api
![GitHub release (latest by date)](https://img.shields.io/github/v/release/jeremmfr/lvslb-api)
[![Go Status](https://github.com/jeremmfr/lvslb-api/workflows/Go%20Tests/badge.svg)](https://github.com/jeremmfr/lvslb-api/actions)
[![Lint Status](https://github.com/jeremmfr/lvslb-api/workflows/GolangCI-Lint/badge.svg)](https://github.com/jeremmfr/lvslb-api/actions)
[![Go Report Card](https://goreportcard.com/badge/github.com/jeremmfr/lvslb-api)](https://goreportcard.com/report/github.com/jeremmfr/lvslb-api)

Create API REST for keepavlived virtual_server (generate file, reload keepalived and ipvs command)

Compile:
--------
export GO111MODULE=on  
go build -o lvslb-api

Run:
----

	 ./lvslb-api -h
	 	Usage of ./lvslb-api:
		-cert string
			file of certificat for https
  		-dir_keepalived string
        		directory for keepalived files (default "/etc/keepalived/keepalived-ipvs.d/")
  		-htpasswd string
        		htpasswd file for login:password
  		-https
        		https = true or false
  		-ip string
        		listen on IP (default "127.0.0.1")
  		-key string
        		file of key for https
  		-log string
        		file for access log (default "/var/log/lvslb-api.access.log")
        -mon_add string
                script for monitoring add witch arguments $protocol $ip $port $backends $mon_period
        -mon_change string
                script for monitoring change witch arguments $protocol $ip $port $backends $mon_period
        -mon_remove string
                script for monitoring remove witch arguments $protocol $ip $port $backends $mon_period
  		-port string
        		listen on port (default "8080")
  		-reload_cmd string
        		command for reload ipvs keepalived process (default "service keepalived reload")
  		-sleep int
        		time for sleep after reload ipvs keepalived (default 5)


***
API List :
---------
**LIST ALL IPVS**
	`/list_ipvs_all/`  
status of all virtual_server ipvs

**LIST ONE IPVS**
	`/list_ipvs/{protocol}/{IP}/{port}/`  
	status for one virtual_server ipvs  
		`?stats=true` for stats instead connections  
		`?zero=true` for reset stats  
		`?count=true` for count real_server in virtual_server pool  

**ADD**
	`/add_ipvs/{proto}/{IP}/{port}/`

**REMOVE**
	`/remove_ipvs/{proto}/{IP}/{port}/`

**CHECK**
	`/check_ipvs/{proto}/{IP}/{port}/`

**MODIFY**
	`/change_ipvs/{proto}/{IP}/{port}/`

All requests need json in body (except LIST) with these parameters :
* **IP** : IP for virtual_server
* **Port** : Port for virtual_server
* **Protocol** : Protocol for virtual_server (TCP|UDP|SCTP)
* **Lb_algo** : Algorithm for load balancing (rr|wrr|lc|wlc|lblc|sh|dh)
* **Lb_kind** : Type of load balancing (NAT|DR|TUN)
* **Persistence_timeout** : Persistence timeout for client sticky to backend
* **Delay_loop** : Timer before each backend check
* **Sorry_IP** : (Optional) IP of sorry server if all backend server is out of pool
* **Sorry_Port** : (Optional) [Default: $Port ] Port of sorry server if all backend server is out of pool
* **[]Backends** : List of backends with this parameters :
	* **IP** : IP of backend
	* **Port** : (Optional) [Default: $Port ] Port of backend
	* **Check_type** : Type of check (TCP_CHECK|HTTP_GET|SSL_GET|MISC_CHECK)
	* **Weight** : (Optional) [Default: 1 ] Weight of backend ratio to total weights
	* **Check_port** : (Optional) [Default: $Backend.Port] Port for check if different of Port
	* **Check_timeout** : (Optional) Timeout for check
	* **Nb_get_retry** : (Optional) Retry before remove backend from the pool
	* **Delay_before_retry** : (Optional) Timer before retry after failure
	* **Url_path** : Url for HTTP_GET and SSL_GET
	* **Url_digest** : (Optional) Digest of response for HTTP_GET and SSL_GET
	* **Url_status_code** : (Optional) HTTP Code of response for HTTP_GET and SSL_GET
	* **Misc_path** : (Optional) Path for MISC_CHECK


CHANGELOG.md
## 1.14
BUG FIXES:
* fix sorry port zero
* fix lint errors

### 1.13

First pub release


file.go
package main

import (
	"fmt"
	"io/ioutil"
	"os"
	"strings"
)

// checkIpvsExists : test if keepalived file exists.
func checkIpvsExists(ipvs ipvsStruc) bool {
	_, err := os.Stat(strings.Join([]string{*dirKeepalived, ipvs.IP, "_", ipvs.Protocol, "_", ipvs.Port, ".conf"}, ""))

	return !os.IsNotExist(err)
}

// generateFile : generate keepalived file string.
func generateFile(ipvs ipvsStruc) string {
	ipvsIn := strings.Join([]string{"virtual_server ", ipvs.IP, " ", ipvs.Port, " {\n"}, "")
	ipvsIn = strings.Join([]string{ipvsIn, "\tprotocol ", ipvs.Protocol,
		"\n\tlb_kind ", ipvs.LbKind,
		"\n\tlb_algo ", ipvs.LbAlgo, "\n"}, "")
	if ipvs.PersistenceTimeout != "0" {
		ipvsIn = strings.Join([]string{ipvsIn, "\tpersistence_timeout ", ipvs.PersistenceTimeout, "\n"}, "")
	}
	ipvsIn = strings.Join([]string{ipvsIn, "\tdelay_loop ", ipvs.DelayLoop, "\n"}, "")
	if ipvs.Virtualhost != "" {
		ipvsIn = strings.Join([]string{ipvsIn, "\tvirtualhost ", ipvs.Virtualhost, "\n"}, "")
	}
	if ipvs.SorryIP != "" {
		ipvsIn = strings.Join([]string{ipvsIn, "\tsorry_server ", ipvs.SorryIP, " "}, "")
		if ipvs.SorryPort != "" && ipvs.SorryPort != "0" {
			ipvsIn = strings.Join([]string{ipvsIn, ipvs.SorryPort, "\n"}, "")
		} else {
			ipvsIn = strings.Join([]string{ipvsIn, ipvs.Port, "\n"}, "")
		}
	}
	for _, backend := range ipvs.Backends {
		ipvsIn = strings.Join([]string{ipvsIn, "\treal_server ", backend.IP, " "}, "")
		if backend.Port != "" {
			ipvsIn = strings.Join([]string{ipvsIn, backend.Port, " {\n"}, "")
		} else {
			ipvsIn = strings.Join([]string{ipvsIn, ipvs.Port, " {\n"}, "")
		}
		if backend.Weight != "" {
			ipvsIn = strings.Join([]string{ipvsIn, "\t\tweight ", backend.Weight, "\n"}, "")
		} else {
			ipvsIn = strings.Join([]string{ipvsIn, "\t\tweight 1\n"}, "")
		}
		if backend.CheckType != checkNONE {
			ipvsIn = strings.Join([]string{ipvsIn, "\t\t", backend.CheckType, " {\n"}, "")
			if backend.CheckType == checkGET || backend.CheckType == checkSSLGET {
				if backend.URLPath != "" {
					ipvsIn = strings.Join([]string{ipvsIn, "\t\t\turl {\n\t\t\t\tpath ", backend.URLPath, "\n"}, "")
					if backend.URLDigest != "" {
						ipvsIn = strings.Join([]string{ipvsIn, "\t\t\t\tdigest ", backend.URLDigest, "\n"}, "")
					}
					if backend.URLStatusCode != "" {
						ipvsIn = strings.Join([]string{ipvsIn, "\t\t\t\tstatus_code ", backend.URLStatusCode, "\n"}, "")
					}
					ipvsIn = strings.Join([]string{ipvsIn, "\t\t\t}\n"}, "")
				}
			}
			if backend.CheckType == checkGET || backend.CheckType == checkSSLGET || backend.CheckType == checkTCP {
				if backend.CheckPort != "" {
					ipvsIn = strings.Join([]string{ipvsIn, "\t\t\tconnect_port ", backend.CheckPort, "\n"}, "")
				}
				if backend.CheckTimeout != "" {
					ipvsIn = strings.Join([]string{ipvsIn, "\t\t\tconnect_timeout ", backend.CheckTimeout, "\n"}, "")
				}
				if backend.NbGetRetry != "" {
					if backend.CheckType == checkGET || backend.CheckType == checkSSLGET {
						ipvsIn = strings.Join([]string{ipvsIn, "\t\t\tnb_get_retry ", backend.NbGetRetry, "\n"}, "")
					}
					if backend.CheckType == checkTCP {
						ipvsIn = strings.Join([]string{ipvsIn, "\t\t\tretry ", backend.NbGetRetry, "\n"}, "")
					}
				}
				if backend.DelayBeforeRetry != "" {
					ipvsIn = strings.Join([]string{ipvsIn, "\t\t\tdelay_before_retry ", backend.DelayBeforeRetry, "\n"}, "")
				}
			}
			if backend.CheckType == checkMISC {
				if backend.MiscPath != "" {
					ipvsIn = strings.Join([]string{ipvsIn, "\t\t\tmisc_path \"", backend.MiscPath, "\"\n"}, "")
				} else {
					ipvsIn = strings.Join([]string{ipvsIn, "\t\t\tmisc_path \"exit 0\"\n"}, "")
				}
				if backend.CheckTimeout != "" {
					ipvsIn = strings.Join([]string{ipvsIn, "\t\t\tmisc_timeout ", backend.CheckTimeout, "\n"}, "")
				}
			}
			ipvsIn = strings.Join([]string{ipvsIn, "\t\t}\n"}, "")
		}
		ipvsIn = strings.Join([]string{ipvsIn, "\t}\n"}, "")
	}
	ipvsIn = strings.Join([]string{ipvsIn, "}\n"}, "")
	if ipvs.MonPeriod != "" && ipvs.MonPeriod != defaultPeriod {
		ipvsIn = strings.Join([]string{ipvsIn, "# mon_period ", ipvs.MonPeriod, "\n"}, "")
	}

	return ipvsIn
}

// checkIpvsOk : compare keepalived file.
func checkIpvsOk(ipvs ipvsStruc) (bool, error) {
	ipvsIn := generateFile(ipvs)

	ipvsReadByte, err := ioutil.ReadFile(strings.Join([]string{*dirKeepalived,
		ipvs.IP, "_", ipvs.Protocol, "_", ipvs.Port, ".conf"},
		""))
	ipvsRead := string(ipvsReadByte)

	if err != nil {
		return false, fmt.Errorf("failed to read file : %w", err)
	}
	if ipvsIn == ipvsRead {
		return true, nil
	}

	return false, nil
}

// addIpvsFile : write generated keepalived file on system.
func addIpvsFile(ipvs ipvsStruc) error {
	ipvsIn := generateFile(ipvs)
	err := ioutil.WriteFile(strings.Join([]string{*dirKeepalived, // nolint: gosec
		ipvs.IP, "_", ipvs.Protocol, "_", ipvs.Port, ".conf"},
		""), []byte(ipvsIn), 0644)
	if err != nil {
		return fmt.Errorf("failed to write file : %w", err)
	}

	return nil
}

// removeIpvsFile : remove keepalived file.
func removeIpvsFile(ipvs ipvsStruc) error {
	err := os.Remove(strings.Join([]string{*dirKeepalived, ipvs.IP, "_", ipvs.Protocol, "_", ipvs.Port, ".conf"}, ""))
	if err != nil {
		return fmt.Errorf("failed to delete file : %w", err)
	}

	return nil
}


ipvs.go
package main

import (
	"fmt"
	"net/http"
	"os/exec"
	"strings"

	"github.com/gorilla/mux"
)

// listIpvsAll : return output of command ipvsadm -L -n.
func listIpvsAll(w http.ResponseWriter, r *http.Request) {
	stdout, err := exec.Command("ipvsadm", "-L", "-n").Output()
	if err != nil {
		http.Error(w, err.Error(), 500)

		return
	}
	fmt.Fprintln(w, string(stdout))
}

// listIpvs : return output of command ipvsadm -L -n or ipvsadm -Z for specific virtual_server.
func listIpvs(w http.ResponseWriter, r *http.Request) {
	vars := mux.Vars(r)
	var command []string

	if r.URL.Query().Get("zero") == trueStr {
		command = append(command, "-Z")
	} else {
		command = append(command, "-L", "-n")
	}

	switch strings.ToLower(vars["proto"]) {
	case "tcp":
		command = append(command, "-t")
	case "udp":
		command = append(command, "-u")
	default:
		w.WriteHeader(http.StatusBadRequest)
		fmt.Fprintln(w, "Unknown Protocol")

		return
	}
	if strings.Contains(vars["IP"], ":") {
		command = append(command, strings.Join([]string{"[", vars["IP"], "]", ":", vars["port"]}, ""))
	} else {
		command = append(command, strings.Join([]string{vars["IP"], ":", vars["port"]}, ""))
	}
	if r.URL.Query().Get("stats") == trueStr && r.URL.Query().Get("zero") == "" {
		command = append(command, "--stats", "--exact")
	}
	stdout, err := exec.Command("ipvsadm", command...).Output()
	if err != nil {
		http.Error(w, err.Error(), 500)

		return
	}
	if r.URL.Query().Get("count") == trueStr {
		realServer := strings.Count(string(stdout), ">")
		if realServer == 0 {
			fmt.Fprintln(w, "Unknown virtual_server")
		} else {
			fmt.Fprintln(w, realServer-1, "real_server(s)")
		}
	}
	fmt.Fprintln(w, string(stdout))
}


main.go
package main

import (
	"flag"
	"log"
	"net/http"
	"os"
	"strings"
	"sync"
	"time"

	"github.com/gorilla/handlers"
	"github.com/gorilla/mux"
)

type ipvsStruc struct {
	IP                 string          `json:"IP"`
	Port               string          `json:"Port"`
	Protocol           string          `json:"Protocol"`
	DelayLoop          string          `json:"Delay_loop"`
	LbAlgo             string          `json:"Lb_algo"`
	LbKind             string          `json:"Lb_kind"`
	PersistenceTimeout string          `json:"Persistence_timeout"`
	SorryIP            string          `json:"Sorry_IP"`
	SorryPort          string          `json:"Sorry_port"`
	Virtualhost        string          `json:"Virtualhost"`
	MonPeriod          string          `json:"Mon_period"`
	Backends           ipvsBackendList `json:"Backends"`
}

type ipvsBackend struct {
	IP               string `json:"IP"`
	Port             string `json:"Port"`
	Weight           string `json:"Weight"`
	CheckType        string `json:"Check_type"`
	CheckPort        string `json:"Check_port"`
	CheckTimeout     string `json:"Check_timeout"`
	NbGetRetry       string `json:"Nb_get_retry"`
	DelayBeforeRetry string `json:"Delay_before_retry"`
	URLPath          string `json:"Url_path"`
	URLDigest        string `json:"Url_digest"`
	URLStatusCode    string `json:"Url_status_code"`
	MiscPath         string `json:"Misc_path"`
}

type ipvsBackendList []ipvsBackend

const checkNONE string = "NONE"
const checkGET string = "HTTP_GET"
const checkSSLGET string = "SSL_GET"
const checkTCP string = "TCP_CHECK"
const checkMISC string = "MISC_CHECK"
const defaultPeriod string = "default"
const trueStr string = "true"

func (slice ipvsBackendList) Len() int {
	return len(slice)
}

func (slice ipvsBackendList) Less(i, j int) bool {
	return slice[i].IP < slice[j].IP
}

func (slice ipvsBackendList) Swap(i, j int) {
	slice[i], slice[j] = slice[j], slice[i]
}

var (
	htpasswdfile            *string
	reloadKeepalivedCommand *string
	dirKeepalived           *string
	timeSleep               *int
	mutex                   = &sync.Mutex{}
	scriptMonitoringAdd     *string
	scriptMonitoringChange  *string
	scriptMonitoringRemove  *string
)

func main() {
	listenIP := flag.String("ip", "127.0.0.1", "listen on IP")
	listenPort := flag.String("port", "8080", "listen on port")
	https := flag.Bool("https", false, "https = true or false")
	cert := flag.String("cert", "", "file of certificat for https")
	key := flag.String("key", "", "file of key for https")
	accessLogFile := flag.String("log", "/var/log/lvslb-api.access.log", "file for access log")
	htpasswdfile = flag.String("htpasswd", "", "htpasswd file for login:password")

	reloadKeepalivedCommand = flag.String("reload_cmd", "service keepalived reload",
		"command for reload ipvs keepalived process")
	dirKeepalived = flag.String("dir_keepalived", "/etc/keepalived/keepalived-ipvs.d/",
		"directory for keepalived files")
	timeSleep = flag.Int("sleep", 5,
		"time for sleep after reload ipvs keepalived")

	scriptMonitoringAdd = flag.String("mon_add", "",
		"script for monitoring add witch arguments $protocol $ip $port $backends $mon_period")
	scriptMonitoringChange = flag.String("mon_change", "",
		"script for monitoring change witch arguments $protocol $ip $port $backends $mon_period")
	scriptMonitoringRemove = flag.String("mon_remove", "",
		"script for monitoring remove witch arguments $protocol $ip $port $backends $mon_period")

	flag.Parse()

	dirinfo, err := os.Stat(*dirKeepalived)
	if err != nil {
		log.Fatalf(strings.Join([]string{"dir_keepalived ", *dirKeepalived, " not exist"}, ""))
	}
	if !dirinfo.IsDir() {
		log.Fatalf(strings.Join([]string{"dir_keepalived ", *dirKeepalived, " not a directory"}, ""))
	}
	*dirKeepalived = strings.TrimSuffix(*dirKeepalived, "/")
	*dirKeepalived = strings.Join([]string{*dirKeepalived, "/"}, "")

	// accesslog file open
	accessLog, err := os.OpenFile(*accessLogFile, os.O_CREATE|os.O_WRONLY|os.O_APPEND, 0644)
	if err != nil {
		log.Fatalf("Failed to open access log: %s", err)
	}

	router := mux.NewRouter().StrictSlash(true)
	router.HandleFunc("/list_ipvs_all/", listIpvsAll)
	router.HandleFunc("/list_ipvs/{proto}/{IP}/{port}/", listIpvs)
	router.HandleFunc("/check_ipvs/{proto}/{IP}/{port}/", checkIpvs)
	router.HandleFunc("/add_ipvs/{proto}/{IP}/{port}/", addIpvs)
	router.HandleFunc("/remove_ipvs/{proto}/{IP}/{port}/", removeIpvs)
	router.HandleFunc("/change_ipvs/{proto}/{IP}/{port}/", changeIpvs)

	loggedRouter := handlers.CombinedLoggingHandler(accessLog, router)
	if *https {
		if (*cert == "") || (*key == "") {
			log.Fatalf("HTTPS true but no cert and key defined")
		} else {
			log.Fatal(http.ListenAndServeTLS(strings.Join([]string{*listenIP, ":", *listenPort}, ""), *cert, *key, loggedRouter))
		}
	} else {
		log.Fatal(http.ListenAndServe(strings.Join([]string{*listenIP, ":", *listenPort}, ""), loggedRouter))
	}
}

func sleep() {
	time.Sleep(time.Duration(*timeSleep) * time.Second)
}
